pkg/agent/controller/egress/egress_controller.go

```diff
@@ -0,0 +1,681 @@
+	c.onEgressGroupRemove(policy.Name)
+	delete(c.setByGroup, policy.Name)
+	return nil
+}
+
+func (c *Controller) ReplaceEgressGroups(groups []*v1beta1.EgressGroup) {
+	c.setByGroupLock.Lock()
+	defer c.setByGroupLock.Unlock()
+
+	oldGroupKeys := make(sets.String, len(c.setByGroup))
+	for key := range c.setByGroup {
+		oldGroupKeys.Insert(key)
+	}
+
+	for _, group := range groups {
+		oldGroupKeys.Delete(group.Name)
+		c.addEgressGroupLocked(group)
+	}
+
+	for key := range oldGroupKeys {
+		delete(c.setByGroup, key)
+	}
+	return
+}
+
+func (c *Controller) DeleteEgressGroup(group *v1beta1.EgressGroup) error {
+	c.setByGroupLock.Lock()
+	defer c.setByGroupLock.Unlock()
+	klog.Infof("%#v", group)
+	return c.DeleteEgressGroupLocked(group)
+}
+
+func (c *Controller) DeleteEgressGroupLocked(group *v1beta1.EgressGroup) error {
+	groupMemberSet := v1beta1.GroupMemberSet{}
+	for i := range group.GroupMembers {
+		groupMemberSet.Insert(&group.GroupMembers[i])
+	}
+	oldGroupMemberSet, exist := c.setByGroup[group.Name]
+	if !exist {
+		return nil
+	}
+	removedMembers := make([]v1beta1.GroupMember, 0, len(group.GroupMembers))
+	NewGroupMemberPodSet := v1beta1.GroupMemberSet{}
+
+	for i := range groupMemberSet {
+		if oldGroupMemberSet.Has(groupMemberSet[i]) {
+			removedMembers = append(removedMembers, *groupMemberSet[i])
+		} else {
+			NewGroupMemberPodSet.Insert(groupMemberSet[i])
+		}
+	}
+	c.setByGroup[group.Name] = NewGroupMemberPodSet
+	c.onEgressRemove(removedMembers)
+	return nil
+}
+
+func (c *Controller) onEgresssAdd(groupName string, members []v1beta1.GroupMember) {
+	ip, exist := c.policyMap[groupName]
+	if !exist {
+		return
+	}
+	for i := range members {
+		c.enqueueRuleAdd(ip, &members[i])
+	}
+}
+
+func (c *Controller) onEgressRemove(members []v1beta1.GroupMember) {
+	for i := range members {
+		c.enqueueRuleDelete(&members[i])
+	}
+}
+
+func (c *Controller) onEgressGroupAdd(groupName string) {
+	ip := c.policyMap[groupName]
+	groups, _ := c.setByGroup[groupName]
+	for _, member := range groups {
+		c.enqueueRuleAdd(ip, member)
+	}
+}
+
+func (c *Controller) onEgressGroupRemove(groupName string) {
+	groups, _ := c.setByGroup[groupName]
+	for _, member := range groups {
+		c.enqueueRuleDelete(member)
+	}
+}
+
+func (c *Controller) addEgressPolicy(obj interface{}) {
+	c.policyMapLock.Lock()
+	defer c.policyMapLock.Unlock()
+
+	egress := obj.(*v1alpha1.Egress)
+	egressIP := egress.Spec.EgressIP
+	klog.Infof(egressIP)
+	if isLocalIP(egressIP) {
+		c.IPAllocatorLock.Lock()
+		defer c.IPAllocatorLock.Unlock()
+		id, err := c.IPAllocator.allocateForIP(egressIP)
+		if err != nil {
+			klog.Errorf(err.Error())
+			return
+		}
+		c.ofClient.InstallSNATMarkFlows(net.ParseIP(egressIP), id)
+		if err != nil {
+			klog.Errorf(err.Error())
+			return
+		}
+		errroute := c.routeClient.AddSNATRule(net.ParseIP(egressIP), id)
+		if errroute != nil {
+			klog.Infof(errroute.Error())
+			return
+		}
+	}
+	c.addEgressPolicyLocked(egress)
```

Comment:
This is not the right pattern to use eventHandler and workqueue:
1. It doesn't have chance to retry if there is any transient error when calling openflow/iptables interfaces.
2. It's not efficient to have blocking calls in the eventHandlers as all events are handled serially. Blocking calls should be in syncHandler of workers so that they can be processed concurrently.